src/pe/authenticode.rs

```diff
@@ -76,24 +96,166 @@ impl<'s> Iterator for ExcludedSectionsIter<'s> {
+                            };
+
+                            return Some(buf);
                         } else {
-                            self.state = IterState::Final(start)
+                            self.state = IterState::Final {
+                                sum_of_bytes_hashed,
+                            };
+                        }
+                    }
+                    IterState::Final {
+                        sum_of_bytes_hashed,
+                    } => {
+                        // 14. Create a value called FILE_SIZE, which is not part of the signature.
+                        //     Set this value to the image’s file size, acquired from the underlying
+                        //     file system. If FILE_SIZE is greater than SUM_OF_BYTES_HASHED, the
+                        //     file contains extra data that must be added to the hash. This data
+                        //     begins at the SUM_OF_BYTES_HASHED file offset, and its length is:
+                        //       (File Size) - ((Size of AttributeCertificateTable) + SUM_OF_BYTES_HASHED)
+                        //
+                        // Note: The size of Attribute Certificate Table is specified in the second
+                        //       ULONG value in the Certificate Table entry (32 bit: offset 132,
+                        //       64 bit: offset 148) in Optional Header Data Directories.
+                        let file_size = bytes.len();
+
+                        // If FILE_SIZE is not a multiple of 8 bytes, the data added to the hash must
+                        // be appended with zero padding of length (8 – (FILE_SIZE % 8)) bytes
+                        let pad_size = (8 - file_size % 8) % 8;
+                        self.state = IterState::Padding(pad_size);
+
+                        if file_size > sum_of_bytes_hashed {
+                            let extra_data_start = sum_of_bytes_hashed;
+                            let len =
+                                file_size - sections.certificate_table_size - sum_of_bytes_hashed;
+
+                            debug!("hashing {extra_data_start:#x} {len:#x}",);
+                            let buf = &bytes[extra_data_start..extra_data_start + len];
+
+                            return Some(buf);
                         }
                     }
-                    IterState::Final(start) => {
+                    IterState::Padding(pad_size) => {
                         self.state = IterState::Done;
-                        return Some(&bytes[start..]);
+
+                        if pad_size != 0 {
+                            debug!("hashing {pad_size:#x}");
+
+                            // NOTE (safety): pad size will be at most 7, and PADDING has a size of 7
+                            //                pad_size is computed ~10 lines above.
+                            debug_assert!(pad_size <= 7);
+                            debug_assert_eq!(PADDING.len(), 7);
```

Comment:
@m4b are those `debug_assert` fine? I thought I'd might as well back up my claims?

Reply:
If they’re some internal check that should only be false if someone refactors this code and doesn’t maintain some internal invariant afterwards then yea. If they’re like things that probably should be true but maybe aren’t if the binary is malformed, then no, should be an error

Reply:
no, they are only meant to protect against a refactor.